tests/grammar/test_grammar.py

```diff
@@ -0,0 +1,129 @@
+}
+
+
+def utf8_encodable(terminal: str) -> bool:
+    try:
+        if '\x00' not in terminal and '\\ ' not in terminal:
+            terminal.encode()
+            return True
+        else:
+            return False
+    except UnicodeEncodeError:  # pragma: no cover
+        # Very rarely, a "." in some terminal regex will generate a surrogate
+        # character that cannot be encoded as UTF-8.  We apply this filter to
+        # ensure it doesn't happen at runtime, but don't worry about coverage.
+        return False
+
+
+class GrammarStrategy(LarkStrategy):
+    def __init__(self, grammar, start, explicit_strategies):
+        super().__init__(grammar, start, explicit_strategies)
+        self.terminal_strategies = {
+            k: v.map(lambda s: s.replace("\0", "")).filter(utf8_encodable)
+            for k, v in self.terminal_strategies.items()  # type: ignore
+        }
+
+    def draw_symbol(self, data, symbol, draw_state):  # type: ignore
+        count = len(draw_state.result)
+        super().draw_symbol(data, symbol, draw_state)
+        if symbol.name in COMPILE_MODES:
+            try:
+                compile(
+                    source="".join(
+                        draw_state.result[count:]
+                    ).replace("contract", "class").replace("struct", "class"),
+                    filename="<string>",
+                    mode=COMPILE_MODES[symbol.name],
+                )
+            except SyntaxError:
+                # Python's grammar doesn't actually fully describe the behaviour of the
+                # CPython parser and AST-post-processor, so we just filter out errors.
+                assume(False)
+
+
+def from_grammar(start: str = "file_input") -> st.SearchStrategy[str]:
+    """Generate syntactically-valid Python source code based on the grammar.
+    Valid values for ``start`` are ``"single_input"``, ``"file_input"``, or
+    ``"eval_input"``; respectively a single interactive statement, a module or
+    sequence of commands read from a file, and input for the eval() function.
+    .. warning::
+        DO NOT EXECUTE CODE GENERATED BY THIS STRATEGY.
+        It could do literally anything that running Python code is able to do,
+        including changing, deleting, or uploading important data.  Arbitrary
+        code can be useful, but "arbitrary code execution" can be very, very bad.
+    """
+    assert start in {"single_input", "file_input", "eval_input"}
+    grammar = get_lark_grammar()
+    explicit_strategies = dict(
+        _INDENT=st.just(" " * 4),
+        _DEDENT=st.just(""),
+        NAME=st.from_regex(r"[a-z_A-Z]+", fullmatch=True).filter(str.isidentifier),
+    )
+    return GrammarStrategy(grammar, start, explicit_strategies)
+
+
+@given(
+    code=from_grammar()
+)
+@hypothesis.settings(
+    deadline=400,
+    max_examples=2000
+)
+def test_grammar_bruteforce(code):
+    if utf8_encodable(code):
+        tree = parser.parse_to_ast(code + "\n")
```

Comment:
Damn Python newline bug!

Reply:
LOL; yes - no biggy IMHO. It's difficult to evaluate module (files) vs. single eval statements.